numba/targets/arraymath.py

```diff
@@ -3541,6 +3542,32 @@ def np_extract_impl(condition, arr):
 
     return np_extract_impl
 
+  
+@overload(np.select)
+def np_select(condlist, choicelist, default=0):
+
+    def np_select_impl(condlist, choicelist, default=0):
+        out = default * np.ones(choicelist[0].shape, choicelist[0].dtype)
+        #should use reversed+zip, but reversed is not available
+        for i in range(len(condlist) - 1, -1, -1):
+            cond = condlist[i]
+            choice = choicelist[i]
+            out = np.where(cond, choice, out)
+        return out
+
+    if ((isinstance(condlist, (types.List, types.UniTuple)))
+        and (isinstance(choicelist, (types.List, types.UniTuple)))
+        and isinstance(condlist[0], types.Array)
+        and isinstance(condlist[0].dtype, types.Boolean)
+        and isinstance(default, (int, types.scalars.Number))):
+        if condlist[0].ndim != choicelist[0].ndim:
```

Comment:
This assumes the elements of the `condlist` container (list or tuple) have an `ndim` attr. Consider a `condlist` composed of two tuples:
```python
x = np.arange(10)
condlist = [tuple(z for z in x), tuple(z**2 for z in x)]
```
control flow would pass all the checks above and then an attribute error would be raised. Also, seems like NumPy is happy with tuples in `condlist`:
```python
import numpy as np
def foo():
    x = np.arange(10)
    a = [x > 2, 3 < x]
    b = [tuple(z for z in x), tuple(z**2 for z in x)]
    return np.select(a, b, default=False)
foo()
```

Reply:
would a `condlist` composed of two tuples pass all the checks in the control flow? I thought this line would stop it: `isinstance(condlist[0], types.Array)`
However, I realize that I did not add a control for `choicelist` elements being arrays. That would fail as you described.
Regarding not allowing tuples inside `condlist`, I did it on purpose but I'll change it if you think we should have them. The reason was that the Numpy documentation describes the inputs as:
**condlist : list of bool ndarrays**
**choicelist : list of ndarrays**
If you think we should match the behaviour rather than the spec, I can add some extra lines to deal with the tuple case.

Reply:
> would a `condlist` composed of two tuples pass all the checks in the control flow? I thought this line would stop it: `isinstance(condlist[0], types.Array)`
> However, I realize that I did not add a control for `choicelist` elements being arrays. That would fail as you described.

Perhaps I meant `choicelist` :)

> Regarding not allowing tuples inside `condlist`, I did it on purpose but I'll change it if you think we should have them. The reason was that the Numpy documentation describes the inputs as:
> **condlist : list of bool ndarrays**
> **choicelist : list of ndarrays**
> If you think we should match the behaviour rather than the spec, I can add some extra lines to deal with the tuple case.

Numba typically tries to replicate NumPy _behaviour_ unless it is unreasonable, in which case issues/fixes are opened against NumPy. The NumPy behaviour here seems reasonable, something array-like is being accepted.

Reply:
@stuartarchibald , adding the tuple case is not so simple as I thought. My implementation of `select` relies on numba's implementation of `where` which does not seem to support tuples (even though Numpy's does). 
Am I missing something?
```
@njit
def foo(cond, x, y):
    return np.where(cond, x, y)

np.where((True, False), (1,2), (3,4))

foo((True, False), (1,2), (3,4))
```

Reply:
@luk-f-a seems like that indeed does not work. I just looked at the Numba source for `np.where`, looks like it could do with an update. I think with view of containing the scope of this PR, perhaps not worry about implementation for e.g. tuples but instead just ensure that type inference fails correctly when they are encountered. How does that sound?

Reply:
sounds good, I'll work on that.